runtime/src/main/java/org/corfudb/runtime/object/VersionLockedObject.java

```diff
@@ -601,7 +601,7 @@ private Object applyUpdateUnsafe(SMREntry entry, long timestamp) {
                 // This entry actually resets the object. So here
                 // we can safely get a new instance, and add the
                 // previous instance to the undo log.
-                entry.setUndoRecord(object.getContext(context));
+                entry.setUndoRecord(object);
```

Comment:
This doesn't look right. The in-memory object itself shouldn't be referenced, instead undo records should be explicitly generated. If the map map's state at v1 is {a,b,c} then a clear happens at v2 and v3 adds {d, e, f}, then undoing the clear would result in an empty map, but it should have restored {a, b, c}. 

Assuming clears are rarely called, you can just set the undo record to false, that would trigger an object rebuild from scratch, which would be correct, but potentially slow.

Reply:
Here are the set of events:

Going Forwards:
```
V1: Add(A, B, C) | Map@{ReferenceX}{A, B, C]
V2: Clear()      | Map@{ReferenceY}{]
                  (Forces a creation of a new object)
                  (Undo record is now pointing at Map@{ReferenceX})
V3: Add(D, E, F) | Map@{ReferenceY}{D, E, F]
```

Going Backwards:
```
From V3 -> V2: Undo Record D, E and F      | Map@{ReferenceY}{]
From V2 -> V1: Restore map at reference X  | Map@{ReferenceX}{A, B, C]
```

What am I missing?

Reply:
The `clear` method doesn't specify an undo function in `CorfuTable`, but the VLO will populate it anyways (special handling), but now I see that the reference is swapped and a new object is created.

I don't think the special handling is needed, since that functionality can be captured with an undofunction, might have been an optimization to eliminate the map copy when generating the undo record.

Reply:
We should definitely look into it since the cost (added complexity) does not give us much benefit. At this point I don't necessarily want to change the behavior, since some of the clients rely on `clear()` to be O(1). If we were to change it, this might case a performance regression on their side.